File: backend/controllers/authController.js
import jwt from 'jsonwebtoken';
import bcrypt from 'bcrypt';
import {
  findUserByIdentifier,
  incrementFailedAttempts,
  resetFailedAttempts,
} from '../models/user.js';

export async function login(req, res) {
  const { identifier, password } = req.body;

  if (!identifier?.trim() || !password) {
    return res.status(400).json({
      success: false,
      error: 'Identifier and password are required',
    });
  }

  try {
    const user = await findUserByIdentifier(identifier.trim());

    if (!user) {
      return res.status(401).json({
        success: false,
        error: 'Invalid username/email or password',
        code: 'INVALID_CREDENTIALS',
      });
    }

    if (user.is_lock) {
      return res.status(403).json({
        success: false,
        error: 'Account locked. Please contact administrator.',
        code: 'ACCOUNT_LOCKED',
      });
    }

    if (user.status !== 'ACTIVE') {
      return res.status(403).json({
        success: false,
        error: 'Account is not active. Please contact administrator.',
        code: 'ACCOUNT_INACTIVE',
      });
    }

    const isPasswordValid = await bcrypt.compare(password, user.password);

    if (!isPasswordValid) {
      const { newAttempts, shouldLock } = await incrementFailedAttempts(
        user.id,
        user.failed_login_attempts ?? 0
      );

      if (shouldLock) {
        return res.status(403).json({
          success: false,
          error: 'Account locked due to too many failed attempts.',
          code: 'ACCOUNT_LOCKED',
        });
      }

      const remaining = 5 - newAttempts;
      return res.status(401).json({
        success: false,
        error: `Invalid username/email or password. ${remaining} attempt${remaining === 1 ? '' : 's'} remaining.`,
        code: 'INVALID_CREDENTIALS',
      });
    }

    await resetFailedAttempts(user.id);

    const token = jwt.sign(
      { userId: user.id, role: user.role },
      process.env.JWT_SECRET,
      { expiresIn: '8h' }
    );

    return res.status(200).json({
      success: true,
      token,
      user: {
        id: user.id,
        username: user.username,
        email: user.email,
        role: user.role,
        department: user.department,
        position: user.position,
      },
    });

  } catch (err) {
    console.error('[Login Error]', err);
    return res.status(500).json({
      success: false,
      error: 'Internal server error',
    });
  }
}

File: backend/controllers/authController.py
from fastapi import Request, HTTPException
from lib.supabase import supabase
import bcrypt
import jwt
import os

from models.user import (
    findUserByIdentifier,
    incrementFailedAttempts,
    resetFailedAttempts,
)

JWT_SECRET = os.getenv("JWT_SECRET")


async def login(request: Request):
    body = await request.json()

    identifier = body.get("identifier")
    password = body.get("password")

    if not identifier or not identifier.strip() or not password:
        raise HTTPException(
            status_code=400,
            detail={
                "success": False,
                "error": "Identifier and password are required",
            },
        )

    try:
        user = await findUserByIdentifier(identifier.strip())

        if not user:
            raise HTTPException(
                status_code=401,
                detail={
                    "success": False,
                    "error": "Invalid username/email or password",
                    "code": "INVALID_CREDENTIALS",
                },
            )

        if user.get("is_lock"):
            raise HTTPException(
                status_code=403,
                detail={
                    "success": False,
                    "error": "Account locked. Please contact administrator.",
                    "code": "ACCOUNT_LOCKED",
                },
            )

        if user.get("status") != "ACTIVE":
            raise HTTPException(
                status_code=403,
                detail={
                    "success": False,
                    "error": "Account is not active. Please contact administrator.",
                    "code": "ACCOUNT_INACTIVE",
                },
            )

        stored_password = user.get("password", "")
        isPasswordValid = False

        if stored_password.startswith("$2b$"):
            isPasswordValid = bcrypt.checkpw(
                password.encode(), stored_password.encode()
            )
        else:
            isPasswordValid = password == stored_password

        if not isPasswordValid:
            result = await incrementFailedAttempts(
                user["id"],
                user.get("failed_login_attempts") or 0,
            )

            newAttempts = result["newAttempts"]
            shouldLock = result["shouldLock"]

            if shouldLock:
                raise HTTPException(
                    status_code=403,
                    detail={
                        "success": False,
                        "error": "Account locked due to too many failed attempts.",
                        "code": "ACCOUNT_LOCKED",
                    },
                )

            remaining = 5 - newAttempts

            raise HTTPException(
                status_code=401,
                detail={
                    "success": False,
                    "error": f"Invalid username/email or password. {remaining} attempt{'' if remaining == 1 else 's'} remaining.",
                    "code": "INVALID_CREDENTIALS",
                },
            )

        await resetFailedAttempts(user["id"])

        token = jwt.encode(
            {"userId": user["id"], "role": user["role"]},
            JWT_SECRET,
            algorithm="HS256",
        )

        return {
            "success": True,
            "token": token,
            "user": {
                "id": user["id"],
                "username": user.get("username"),
                "email": user.get("email"),
                "role": user.get("role"),
                "department": user.get("department"),
                "position": user.get("position"),
            },
        }

    except HTTPException:
        raise
    except Exception as err:
        print("[Login Error]", err)
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Internal server error",
            },
        )


def authenticate(request: Request):
    auth_header = request.headers.get("authorization")

    if not auth_header or not auth_header.startswith("Bearer "):
        raise HTTPException(
            status_code=401,
            detail={
                "success": False,
                "error": "Authentication required. Please login.",
                "code": "AUTH_REQUIRED",
            },
        )

    token = auth_header.split(" ")[1]

    try:
        decoded = jwt.decode(token, JWT_SECRET, algorithms=["HS256"])
        request.state.user = decoded
        return decoded
    except Exception as err:
        print("[Auth Middleware Error]", str(err))

        error_message = (
            "Token expired"
            if err.__class__.__name__ == "ExpiredSignatureError"
            else "Invalid token"
        )

        raise HTTPException(
            status_code=401,
            detail={
                "success": False,
                "error": error_message,
                "code": "INVALID_TOKEN",
            },
        )